### PR TITLE
HPCC-17957 Report overlapping include paths as an error

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Tested with various command lines to ensure it is reported correctly, and that --checkIncludePaths- disables the checking.

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.  Note --checkIncludePaths is deliberately not included in the help.

File: ecl/eclcc/eclcc.hpp [hunk 1]
```diff
@@ -94,13 +94,14 @@ const char * const helpText[] = {
...
     "!   --logdetail=n Set the level of detail in the log file",
+    "!   --maxErrors=<n> Limit the number of errors, aborting on the nth (default = 5)",
     "!   --nologfile   Do not write any logfile",
 #ifdef _WIN32
     "!   -m            Enable leak checking",
 #endif
     "!   --nogpg       Do not run gpg to check signatures on signed code",
     "    --nosourcepath Compile as if the source came from stdin",
-    "!   --maxErrors=<n> Limit the number of errors, aborting on the nth (default = 5)",
+    "!   --nostdinc    Do not include the current directory in -I",
```
Comment (shamser) [inline ecl/eclcc/eclcc.hpp]: @ghalliday Has the help for maxErrors option been removed deliberately?

File: ecl/eclcc/eclcc.cpp [hunk 2]
```diff
@@ -1791,6 +1792,42 @@ bool EclCC::generatePrecompiledHeader()
...
+    for (unsigned i=1; i < expandedPaths.ordinality(); i++)
+    {
+        const char * prev = expandedPaths.item(i-1);
+        const char * next = expandedPaths.item(i);
+        if (hasPrefix(next, prev, true))
+        {
+            if (!streq(next, prev))
+                throw MakeStringException(99, "Include paths -I '%s' and '%s' overlap", prev, next);
+            else
+                throw MakeStringException(99, "Include path -I '%s' duplicated", prev);
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp]: @ghalliday is it necessary to throw and error for duplicates?

**Comment by ghalliday** [comment]
Repushed to include a fix for the regression test engine.

File: ecl/eclcc/eclcc.cpp [hunk 3]
```diff
@@ -1791,6 +1792,42 @@ bool EclCC::generatePrecompiledHeader()
...
+        if (*cur)
+        {
+            StringBuffer expanded;
+            makeAbsolutePath(cur, expanded);
+            expandedPaths.append(expanded);
+        }
+    }
+
+    //Sort alphabetically, smallest strings will come first
+    expandedPaths.sortAscii(false);
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp]: @ghalliday Could do a case insensitive sorts here for Windows binary.

**Review by shamser — COMMENTED** [review]
@ghalliday Code reviewed.. There's a few comments.

File: ecl/eclcc/eclcc.hpp (hunk 1 above)
Comment (ghalliday) [inline ecl/eclcc/eclcc.hpp]: it was moved so it was in alphabetical order.

File: ecl/eclcc/eclcc.cpp (hunk 2 above)
Comment (ghalliday) [inline ecl/eclcc/eclcc.cpp]: no, but it is inefficient.  I'll remove.

**Comment by ghalliday** [comment]
@shamser PR updated

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.  Looks good.

**Comment by richardkchapman** [comment]
@ghalliday Did Vlad work out why he was providing overlapping paths and how he could stop doing so?

**Comment by ghalliday** [comment]
@richardkchapman He was specifying -I ./subdir - which will have clashed with the default -I ..  However I am not sure that it was the whole story.

**Pull request merged**